Add ability to disable UDP hole punching

GrayHatter opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/c-toxcore/262)
<!-- Reviewable:end -->

GrayHatter commented:
closes #254

endoffile78 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: 0 of 14 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-:-KWKUMA6nIaVN-9uggwB:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 0 of 14 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[auto_tests/dht_test.c, line 306 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKSi8w1R00RmgP3bFp:-KWKSi8w1R00RmgP3bFq:bsld9zd) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/auto_tests/dht_test.c#L306)):*
> ```C
>     ck_assert_msg(net != 0, "Failed to create Networking_Core");
> 
>     DHT *dht = new_DHT(NULL, net. 2);
> ```

2 is not a member of net. You meant `,`.

---

*[auto_tests/dht_test.c, line 443 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKSm2DupZV4jLqAGd_:-KWKSm2DupZV4jLqAGda:b-7dh1bx) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/auto_tests/dht_test.c#L443)):*
> ```C
>         ip_init(&ip, 1);
> 
>         dhts[i] = new_DHT(NULL, new_networking(NULL, ip, DHT_DEFAULT_PORT + i ), 2);
> ```

Did this run through astyle?

---

*[other/bootstrap_daemon/src/tox-bootstrapd.c, line 246 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWYsJIeDkO9LnnDMT:-KWKWYsJIeDkO9LnnDMU:b6kblek) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/other/bootstrap_daemon/src/tox-bootstrapd.c#L246)):*
> ```C
>     }
> 
>     DHT *dht = new_DHT(NULL, net, 2);
> ```

You can include tox.h and use the symbolic constant for NORMAL here and in other places where you added a literal 2.

---

*[toxcore/DHT.c, line 2093 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWqmX0KTFXchRZAJm:-KWKWqmX0KTFXchRZAJn:bg2ktr7) ([raw file](https://github.com/toktok/c-toxcore/blob/f08310794f89dda9bac236a37898b1dedaa1b1b1/toxcore/DHT.c#L2093)):*
> ```C
>     }
> 
>     if (!dht->nat_hole_punching_level) {
> ```

I suggest using `== 0` explicitly, because later we'll have `== 1` and others (or named constants).

---

*[toxcore/DHT.h, line 241 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKUOVvY7GcZpSjlbR2:-KWKUOVwAa5_Gi5k07kx:bixsqjg) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/DHT.h#L241)):*
> ```objc
>     Networking_Core *net;
> 
>     uint8_t     nat_hole_punching_level;
> ```

Align with other members or remove alignment altogether.

---

*[toxcore/tox.api.h, line 367 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKUgX-jJ5wJ8t9Zpzl:-KWKUgX0i49e-5v_9FCn:b-5i2lzt) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L367)):*
> ```objc
>   /**
>    * Only try small port ranges, and wait a long time between attempts.
>    * (This Level Currently Unsupported; defaults to NORMAL)
> ```

No need to capitalise. I suggest: `This level is currently unsupported and is equivalent to $NORMAL.` without parentheses.

---

*[toxcore/tox.api.h, line 371 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKUpLXh3Sq6ORW6iaW:-KWKUpLXh3Sq6ORW6iaX:bboreeo) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L371)):*
> ```objc
>   POLITE,
>   /**
>    * Try to establing a direct UDP connection through any supported NAT device.
> ```

* "establish"
* This should probably mention again that it's doing that using UDP hole-punching.

---

*[toxcore/tox.api.h, line 376 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKUzWWFzAxXLl2fAVy:-KWKUzWWFzAxXLl2fAVz:b-hxqivg) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L376)):*
> ```objc
>   NORMAL,
>   /**
>    * Agressivly try to establish a direct connection to all friends.
> ```

"Aggressively"

---

*[toxcore/tox.api.h, line 377 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKV1IhGGMi8EBToD8q:-KWKV1IhGGMi8EBToD8r:b8x07kh) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L377)):*
> ```objc
>   /**
>    * Agressivly try to establish a direct connection to all friends.
>    * USE WITH CAUTION, this may interferr with some poorly configured NAT devices
> ```

"interfere"

---

*[toxcore/tox.api.h, line 379 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKV6X3zk9bp5p4TX1c:-KWKV6X3zk9bp5p4TX1d:b-q64pkt) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L379)):*
> ```objc
>    * USE WITH CAUTION, this may interferr with some poorly configured NAT devices
>    * leading to a denial of service effect to all devices behind the NAT.
>    * (This Level Currently Unsupported; defaults to NORMAL)
> ```

See above.

---

*[toxcore/tox.api.h, line 381 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKV9Z_yAis7q4Lbd1b:-KWKV9Z_yAis7q4Lbd1c:b3g1knd) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L381)):*
> ```objc
>    * (This Level Currently Unsupported; defaults to NORMAL)
>    */
>   AGRESSIVE,
> ```

AGGRESSIVE

---

*[toxcore/tox.api.h, line 527 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWNiUvBJpMTIwMzbB:-KWKWNiUvBJpMTIwMzbC:bv2k85g) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L527)):*
> ```objc
>        * and end_port were swapped.
>        */
>       uint16_t start_port;
> ```

Is this only DHT? I thought this port is also used for all other communications (`net_crypto`).

---

*[toxcore/tox.api.h, line 548 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKVBR1AaQaSCoqO6Xm:-KWKVBR1AaQaSCoqO6Xn:b-xwqm03) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L548)):*
> ```objc
> 
>       /**
>        * How agressive toxcore will be when trying to establish a direct UDP
> ```

I suggest: "How aggressively toxcore will try to [...]". On the first read I got a bit confused. If others disagree, fix the typo: "aggressive".

---

*[toxcore/tox.api.h, line 551 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKVU0h8yGNYN494JfE:-KWKVU0h8yGNYN494JfF:b-jrj4th) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L551)):*
> ```objc
>        * connection with friends through a Network Address Translation device.
>        *
>        * Care must be used when changing this from the default as too low a level
> ```

"Care must be taken"

However, this sentence is too fancy for technical documentation. I don't think you need to tell people to take care if you tell them the effects of low/high levels. They will know what the effects are, so they will know that they need to take care.

---

*[toxcore/tox.api.h, line 553 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWH2o8Vzns9Q2JnNa:-KWKWH2o8Vzns9Q2JnNb:b-8smvu1) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L553)):*
> ```objc
>        * Care must be used when changing this from the default as too low a level
>        * could stop toxcore from creating a direct connection to friends, and too
>        * high a level could cause a DOS effect, and disrupt ALL network traffic for
> ```

"DoS"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Reviewed 13 of 14 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[auto_tests/dht_test.c, line 306 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKSi8w1R00RmgP3bFp:-KWKfCG8FgCyLYe-QxUJ:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/auto_tests/dht_test.c#L306)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> 2 is not a member of net. You meant `,`.
</details>
Done.

---

*[auto_tests/dht_test.c, line 443 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKSm2DupZV4jLqAGd_:-KWKfIiDY1mudY2GXm8v:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/auto_tests/dht_test.c#L443)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Did this run through astyle?
</details>
Done.

---

*[other/bootstrap_daemon/src/tox-bootstrapd.c, line 246 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWYsJIeDkO9LnnDMT:-KWKfKEqGbt1mmstnGiw:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/other/bootstrap_daemon/src/tox-bootstrapd.c#L246)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> You can include tox.h and use the symbolic constant for NORMAL here and in other places where you added a literal 2.
</details>
Done.

---

*[toxcore/DHT.c, line 2093 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWqmX0KTFXchRZAJm:-KWKgzSW2tXnYDtjU_6O:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/f08310794f89dda9bac236a37898b1dedaa1b1b1/toxcore/DHT.c#L2093)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> I suggest using `== 0` explicitly, because later we'll have `== 1` and others (or named constants).
</details>
Done.

---

*[toxcore/DHT.h, line 241 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKUOVvY7GcZpSjlbR2:-KWKh-MIv8R5q69H_ccP:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/DHT.h#L241)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Align with other members or remove alignment altogether.
</details>
Done.

---

*[toxcore/tox.api.h, line 367 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKUgX-jJ5wJ8t9Zpzl:-KWKh04I98Px57SAHQZC:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L367)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> No need to capitalise. I suggest: `This level is currently unsupported and is equivalent to $NORMAL.` without parentheses.
</details>
Done.

---

*[toxcore/tox.api.h, line 371 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKUpLXh3Sq6ORW6iaW:-KWKh0T1BmMkXArIki3k:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L371)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> * "establish"
> * This should probably mention again that it's doing that using UDP hole-punching.
</details>
Done.

---

*[toxcore/tox.api.h, line 376 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKUzWWFzAxXLl2fAVy:-KWKh0znOQTde_heUFZb:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L376)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> "Aggressively"
</details>
Done.

---

*[toxcore/tox.api.h, line 377 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKV1IhGGMi8EBToD8q:-KWKh1HxU5G7dBunUUWE:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L377)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> "interfere"
</details>
Done.

---

*[toxcore/tox.api.h, line 379 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKV6X3zk9bp5p4TX1c:-KWKh1a_6FV0XafYrvCP:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L379)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> See above.
</details>
Done.

---

*[toxcore/tox.api.h, line 381 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKV9Z_yAis7q4Lbd1b:-KWKh1uB8Xu4-Iw7kwlq:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L381)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> AGGRESSIVE
</details>
Done.

---

*[toxcore/tox.api.h, line 527 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWNiUvBJpMTIwMzbB:-KWKh2JbXwBJlwYlbPCR:b7iz6x8) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L527)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Is this only DHT? I thought this port is also used for all other communications (`net_crypto`).
</details>
It COULD be both, but the other is IMO misleading, as everything having to do with ports is started, or handled by the DHT. Yes Networking_Core is technically lower, but this way, if we change something later, we don't have to break the api to add something that uses separate ports from the DHT.

---

*[toxcore/tox.api.h, line 548 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKVBR1AaQaSCoqO6Xm:-KWKiDfaqZZn5x9tr8Ui:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L548)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> I suggest: "How aggressively toxcore will try to [...]". On the first read I got a bit confused. If others disagree, fix the typo: "aggressive".
</details>
Done.

---

*[toxcore/tox.api.h, line 551 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKVU0h8yGNYN494JfE:-KWKiEEAUJiy3QaX9q1j:b8irsb5) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L551)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> "Care must be taken"
> 
> However, this sentence is too fancy for technical documentation. I don't think you need to tell people to take care if you tell them the effects of low/high levels. They will know what the effects are, so they will know that they need to take care.
</details>
rewirtten

---

*[toxcore/tox.api.h, line 553 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWH2o8Vzns9Q2JnNa:-KWKiFaZavN9ZcSMCMbG:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L553)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> "DoS"
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[auto_tests/dht_test.c, line 306 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKk4z35Jvu2zJJR7wR:-KWKk4z35Jvu2zJJR7wS:b1jljh5) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/auto_tests/dht_test.c#L306)):*
> ```C
>     ck_assert_msg(net != 0, "Failed to create Networking_Core");
> 
>     DHT *dht = new_DHT(NULL, net, 2);
> ```

Tests can include tox.h. Use it to get the NORMAL constant here (and in other places).

---

*[toxcore/DHT.c, line 2093 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKksnadPgAS7gktxZK:-KWKksnadPgAS7gktxZL:b-5ilewq) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/DHT.c#L2093)):*
> ```C
>     }
> 
>     // TODO(toktok/c-toxcore#214)
> ```

I'd like a short statement about what is to do here. Given that `nat_hole_punching_level` is a uint8_t, it's not immediately clear what should be done here.

---

*[toxcore/DHT.c, line 2094 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKl6S3XYqtNeshpnh1:-KWKl6S3XYqtNeshpnh2:b-9zg1cu) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/DHT.c#L2094)):*
> ```C
> 
>     // TODO(toktok/c-toxcore#214)
>     switch (dht->nat_hole_punching_level) {
> ```

Isn't this check a bit too late? `do_NAT` executes and (maybe?) sends NAT pings and does other stuff that it doesn't need to do when hole punching is disabled. You may want an additional `== 0` check in there to completely disable any `do_NAT` activity.

Do check whether this suggestion makes sense. I haven't checked it thoroughly.

---

*[toxcore/tox.api.h, line 551 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKnStBd7Cdd7x8IVBG:-KWKnStBd7Cdd7x8IVBH:b-4rctti) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/tox.api.h#L551)):*
> ```objc
>       /**
>        * Sets how aggressive toxcore will be when trying to establish a direct UDP
>        * connection "hole punching" with friends when behind a NAT device.
> ```

I think some sentence merge conflict happened here.

---

*[toxcore/tox.api.h, line 553 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKnYhjCAgnco128nLp:-KWKnYhjCAgnco128nLq:b-8rz4uh) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/tox.api.h#L553)):*
> ```objc
>        * connection "hole punching" with friends when behind a NAT device.
>        *
>        * Clients should be aware of the implications of changing this from it's default.
> ```

"its"

---

*[toxcore/tox.api.h, line 554 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKndJlJGLo0E8s9avi:-KWKndJlJGLo0E8s9avj:b-bjdfzd) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/tox.api.h#L554)):*
> ```objc
>        *
>        * Clients should be aware of the implications of changing this from it's default.
>        * Too low a level could stop toxcore from creating a direct connections, and too
> ```

"Setting it too low could [...]".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[auto_tests/dht_test.c, line 306 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKk4z35Jvu2zJJR7wR:-KWQW1pwfsDKzmL3aM8u:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/auto_tests/dht_test.c#L306)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Tests can include tox.h. Use it to get the NORMAL constant here (and in other places).
</details>
Done.

---

*[toxcore/DHT.c, line 2093 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKksnadPgAS7gktxZK:-KWQXCAG7x3vLx-4ZcO_:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/DHT.c#L2093)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> I'd like a short statement about what is to do here. Given that `nat_hole_punching_level` is a uint8_t, it's not immediately clear what should be done here.
</details>
Done.

---

*[toxcore/DHT.c, line 2094 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKl6S3XYqtNeshpnh1:-KWQXDRsyW-B2vU0YO8P:b-a964qh) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/DHT.c#L2094)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Isn't this check a bit too late? `do_NAT` executes and (maybe?) sends NAT pings and does other stuff that it doesn't need to do when hole punching is disabled. You may want an additional `== 0` check in there to completely disable any `do_NAT` activity.
> 
> Do check whether this suggestion makes sense. I haven't checked it thoroughly.
</details>
I could, and it would make sense in this change, but I don't think I should. Toxcore will first try to connect on the port suggested by other nodes who say they're already connected to the target peer. 

This change won't stop toxcore from trying to connect via UDP, it only stops it from punching extra holes. Which is what we want.

If you want to disable ANY kind of DHT connections, you can just disable UDP which will give the effect in your first comment.

---

*[toxcore/tox.api.h, line 551 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKnStBd7Cdd7x8IVBG:-KWQYa5CFgh90fNuUcKu:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/tox.api.h#L551)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> I think some sentence merge conflict happened here.
</details>
Done.

---

*[toxcore/tox.api.h, line 553 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKnYhjCAgnco128nLp:-KWQYmkZzfRDNYUXTuAq:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/tox.api.h#L553)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> "its"
</details>
Done.

---

*[toxcore/tox.api.h, line 554 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKndJlJGLo0E8s9avi:-KWQYn5LGVZYkS10FLBR:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/6bcf76f67288f5beab760bc25093037a1dac0dc6/toxcore/tox.api.h#L554)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> "Setting it too low could [...]".
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
This pull is reliably eating toxcore profiles. in the uTox development branch.

Can someone else confirm?

Also, backup any profile you care about before trying

nurupo commented:
So, I assume this won't get merged because of profile eating?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-:-KWSNKEVf7On5WCyXziq:b-wd126)*
<!-- Sent from Reviewable.io -->

nurupo commented:
This PR doesn't touch any savedata code, how it can cause profile eating?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-:-KWSOB9K_E0qkCNsSOiA:b53hsfs)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Anyway, regarding the changed in the PR.

Why the user of toxcore should be able to specify hole punching level? Hole punching rate really depends on NATs of both the sender and the receiver, so setting it low might make it so that you lose connection with some friends constantly. This is really something toxcore should be handling internally, transparently from the user. Letting users set low holepunching just because their router dies and having friends constantly going online/offline because of this is probably bad idea.

Also, see https://github.com/TokTok/c-toxcore/issues/254#issuecomment-260180288

---

Reviewed 7 of 14 files at r1, 1 of 1 files at r2, 2 of 6 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[auto_tests/onion_test.c, line 148 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSPncnycHF8rZ2fbTQ:-KWSPncnycHF8rZ2fbTR:b32vbm) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/auto_tests/onion_test.c#L148)):*
> ```C
>     ip.ip6.uint8[15] = 1;
>     Onion *onion1 = new_onion(new_DHT(NULL, new_networking(NULL, ip, 34567), 2));
>     Onion *onion2 = new_onion(new_DHT(NULL, new_networking(NULL, ip, 34568), 2));
> ```

2?

---

*[auto_tests/onion_test.c, line 225 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSPn3QAUg_ZYMY--l6:-KWSPn3QAUg_ZYMY--l7:b32vbm) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/auto_tests/onion_test.c#L225)):*
> ```C
> 
>     c_sleep(1000);
>     Onion *onion3 = new_onion(new_DHT(NULL, new_networking(NULL, ip, 34569), 2));
> ```

2?

---

*[auto_tests/onion_test.c, line 288 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSPmCnAfu2v8ROketf:-KWSPmCnAfu2v8ROketg:b32vbm) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/auto_tests/onion_test.c#L288)):*
> ```C
>     ip.ip6.uint8[15] = 1;
>     Onions *on = (Onions *)malloc(sizeof(Onions));
>     DHT *dht = new_DHT(NULL, new_networking(NULL, ip, port), 2);
> ```

2?

---

*[other/DHT_bootstrap.c, line 123 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSPkQpS07lntEM16ZD:-KWSPkQpS07lntEM16ZE:b32vbm) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/other/DHT_bootstrap.c#L123)):*
> ```C
>     ip_init(&ip, ipv6enabled);
> 
>     DHT *dht = new_DHT(NULL, new_networking(NULL, ip, PORT), 2);
> ```

2?

---

*[testing/DHT_test.c, line 196 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSTlaoLB37bP0GkoVY:-KWSTlaoLB37bP0GkoVZ:b32vbm) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/testing/DHT_test.c#L196)):*
> ```C
>     ip_init(&ip, ipv6enabled);
> 
>     DHT *dht = new_DHT(NULL, new_networking(NULL, ip, PORT), 2);
> ```

2?

---

*[toxcore/tox.api.h, line 527 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWNiUvBJpMTIwMzbB:-KWSS6KQi4N641P34dX0:bcfmgk8) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L527)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary>
> It COULD be both, but the other is IMO misleading, as everything having to do with ports is started, or handled by the DHT. Yes Networking_Core is technically lower, but this way, if we change something later, we don't have to break the api to add something that uses separate ports from the DHT.
</details>
I don't understand this reason. The ports are used for all the communication, not DHT only.

---

*[toxcore/tox.h, line 586 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSRP8LnQ4Bu86RFweR:-KWSRP8LnQ4Bu86RFweS:bc5a2zy) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/toxcore/tox.h#L586)):*
> ```objc
>      * and end_port were swapped.
>      */
>     uint16_t dht_start_port;
> ```

Why this rename? This port is used not only for DHT.

---

*[toxcore/tox.h, line 592 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSRhJMfkYrA_hBjAeZ:-KWSRhJMfkYrA_hBjAe_:bc5a2zy) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/toxcore/tox.h#L592)):*
> ```objc
>      * The end port of the inclusive port range to attempt to use.
>      */
>     uint16_t dht_end_port;
> ```

Why this rename? This port is used not only for DHT.

---

*[toxcore/tox.h, line 606 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSRi7am60VrSxMdB3C:-KWSRi7am60VrSxMdB3D:bc5a2zy) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/toxcore/tox.h#L606)):*
> ```objc
>      * to disable it.
>      */
>     uint16_t dht_tcp_port;
> ```

Why this rename? This port is used not only for DHT.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-:-KWSU3n2873NOYpr2khV:bxwa8ey)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
It was because something was seriously wrong with my installed headers. Funny when I want the system to find headers... no problem, when I don't... ERRORs

I think this is good to merge pending review approval.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[auto_tests/onion_test.c, line 148 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSPncnycHF8rZ2fbTQ:-KWURzhZg1I3-msDD432:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/auto_tests/onion_test.c#L148)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> 2?
</details>
Done.

---

*[auto_tests/onion_test.c, line 225 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSPn3QAUg_ZYMY--l6:-KWUS-AIaLF9S1ZCi4EE:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/auto_tests/onion_test.c#L225)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> 2?
</details>
Done.

---

*[auto_tests/onion_test.c, line 288 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSPmCnAfu2v8ROketf:-KWUS-XXonNHuBOGkPLI:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/auto_tests/onion_test.c#L288)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> 2?
</details>
Done.

---

*[other/DHT_bootstrap.c, line 123 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSPkQpS07lntEM16ZD:-KWUS05breAUbXF1Uxmw:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/other/DHT_bootstrap.c#L123)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> 2?
</details>
Done.

---

*[testing/DHT_test.c, line 196 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSTlaoLB37bP0GkoVY:-KWUSGic4imgbGUabsFF:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/testing/DHT_test.c#L196)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> 2?
</details>
Done.

---

*[toxcore/tox.api.h, line 527 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWNiUvBJpMTIwMzbB:-KWUSIBu_eUC4nh8dPvZ:b8b84mu) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L527)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> I don't understand this reason. The ports are used for all the communication, not DHT only.
</details>
Another reason is because when you disable UDP, this setting really has no effect at all. I still think renaming it to DHT_[port] is more clear about what it's really effecting, but I'm willing to revert if no one else does.

---

*[toxcore/tox.h, line 606 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSRi7am60VrSxMdB3C:-KWUSjKlocoRYNvf8qJx:b-hw8dda) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/toxcore/tox.h#L606)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Why this rename? This port is used not only for DHT.
</details>
Per my previous comment, this one is still used when UDP is disabled, so another option is to revert just this one.

Or we could rename it to tcp_server/tcp_listen

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-:-KWURVmQUwqKstD8NJjS:b-12hkc8)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[toxcore/tox.api.h, line 527 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWNiUvBJpMTIwMzbB:-KWUu5a4NZv7vK_XSnhk:b9qhnni) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L527)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary>
> Another reason is because when you disable UDP, this setting really has no effect at all. I still think renaming it to DHT_[port] is more clear about what it's really effecting, but I'm willing to revert if no one else does.
</details>
If you disable UDP, doesn't toxcore use start_port and end_port for the TCP socket?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[toxcore/tox.api.h, line 527 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWNiUvBJpMTIwMzbB:-KWUuZ-4eC4PRfLJvWEb:b-4wr6v5) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L527)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> If you disable UDP, doesn't toxcore use start_port and end_port for the TCP socket?
</details>
Hm, I see it doesn't.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

nurupo commented:
*[toxcore/tox.api.h, line 527 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWNiUvBJpMTIwMzbB:-KWUuiR6yp9J9f7kj7z7:b4gwv2o) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L527)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Hm, I see it doesn't.
</details>
It makes more sense to rename it to `udp_start_port` then. That would require renaming the `tcp_port` to `relay_server_tcp_port`, because keeping it as `tcp_port` might make people think it used for the same thing as the `udp_start_port`/`udp_end_port` but just using different protocol, which would be incorrect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 1 of 2 files at r5.
Review status: 13 of 14 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Pushing to v0.0.5 because this exceeds the scope of the bug, which was to add an option to disable hole punching, while this PR introduces aggressiveness levels and breaks the API by renaming fields in Tox_Options.

GrayHatter commented:
I reverted a lot of the changes, this pull in simply an addition of the bool, hole_punching enabled, disabled

---

Review status: 1 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[toxcore/tox.api.h, line 527 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWKWNiUvBJpMTIwMzbB:-KWlO--zD4P7fdp_1Bus:bsfg2uf) ([raw file](https://github.com/toktok/c-toxcore/blob/e973629923475bd2f5ebda7c039b5a9658da2602/toxcore/tox.api.h#L527)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> It makes more sense to rename it to `udp_start_port` then. That would require renaming the `tcp_port` to `relay_server_tcp_port`, because keeping it as `tcp_port` might make people think it used for the same thing as the `udp_start_port`/`udp_end_port` but just using different protocol, which would be incorrect.
</details>
reverted to a simple pull

---

*[toxcore/tox.h, line 586 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSRP8LnQ4Bu86RFweR:-KWlO9mDDvkFziqNCx3H:bvth56w) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/toxcore/tox.h#L586)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Why this rename? This port is used not only for DHT.
</details>
reverted

---

*[toxcore/tox.h, line 592 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWSRhJMfkYrA_hBjAeZ:-KWlOAaY0XE25qNXCqdZ:bvth56w) ([raw file](https://github.com/toktok/c-toxcore/blob/95b4637f739498d2e134f12025f31c120c41ff22/toxcore/tox.h#L592)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Why this rename? This port is used not only for DHT.
</details>
reverted

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-:-KWlOSPCKIrk1QJnfpB9:b-hkwo9n)*
<!-- Sent from Reviewable.io -->

zetok commented:
Review status: 1 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[toxcore/tox.api.h, line 524 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWw9XFtjn3xu7pL3QwN:-KWw9XFuoDEFCFFYAFti:b-y56iir) ([raw file](https://github.com/toktok/c-toxcore/blob/b3eca744e07fd4b0c7ca5aef222c0f7017e25631/toxcore/tox.api.h#L524)):*
> ```objc
>      * Enables or disables UDP hole-punching in toxcore, (Default: enabled).
>      */
>     bool nat_level;
> ```

It's not clear what the bool is about just from the bool name. Shouldn't it be renamed to something else?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Reviewed 11 of 13 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[toxcore/tox.api.h, line 522 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KX-aBa1Z-sUi-Rzz4MF:-KX-aBa1Z-sUi-Rzz4MG:b-jg1jga) ([raw file](https://github.com/toktok/c-toxcore/blob/b3eca744e07fd4b0c7ca5aef222c0f7017e25631/toxcore/tox.api.h#L522)):*
> ```objc
> 
>     /**
>      * Enables or disables UDP hole-punching in toxcore, (Default: enabled).
> ```

Should we be more explicit about how the true/false maps to "enables"/"disables"? Reading this I could guess that setting `nat_level = true` would enable hole-punching, but I don't like to guess.

---

*[toxcore/tox.api.h, line 524 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWw9XFtjn3xu7pL3QwN:-KX-a6_ORZHmao8ALCnM:b-dq0qw5) ([raw file](https://github.com/toktok/c-toxcore/blob/b3eca744e07fd4b0c7ca5aef222c0f7017e25631/toxcore/tox.api.h#L524)):*
<details><summary><i>Previously, zetok wrote…</i></summary>
> It's not clear what the bool is about just from the bool name. Shouldn't it be renamed to something else?
</details>
`nat_level` is a weird name for disabling/enabling hole punching. What is a "nat level" anyway?`nat_holepunching_enabled` or `nat_holepunching` might be better.

Also, are we sure that we won't ever want to extend this to hold more than 2 values (true/false)? Using an enum might be more preferable API-compatibility-wise if we decide to extend this in the future. As a bonus, even if we never get to extend it, it makes the client code more readable, e.g. `new_DHT(NULL, net, TOX_NAT_HOLEPUNCHING_NORMAL);` vs `new_DHT(NULL, net, true);`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[toxcore/tox.api.h, line 522 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KX1HmdtYaneO-qO-qi4:-KX1HmdtYaneO-qO-qi5:b6re0rz) ([raw file](https://github.com/toktok/c-toxcore/blob/b3eca744e07fd4b0c7ca5aef222c0f7017e25631/toxcore/tox.api.h#L522)):*
> ```objc
> 
>     /**
>      * Enables or disables UDP hole-punching in toxcore, (Default: enabled).
> ```

Replace comma with period.

---

*[toxcore/tox.api.h, line 524 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWw9XFtjn3xu7pL3QwN:-KX1Hpu77E41JpmSFQu_:b-m2d7cz) ([raw file](https://github.com/toktok/c-toxcore/blob/b3eca744e07fd4b0c7ca5aef222c0f7017e25631/toxcore/tox.api.h#L524)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> `nat_level` is a weird name for disabling/enabling hole punching. What is a "nat level" anyway?`nat_holepunching_enabled` or `nat_holepunching` might be better.
> 
> Also, are we sure that we won't ever want to extend this to hold more than 2 values (true/false)? Using an enum might be more preferable API-compatibility-wise if we decide to extend this in the future. As a bonus, even if we never get to extend it, it makes the client code more readable, e.g. `new_DHT(NULL, net, TOX_NAT_HOLEPUNCHING_NORMAL);` vs `new_DHT(NULL, net, true);`.
</details>
For now, we're sure, but anyway this bool is going away very soon, when it becomes part of the "nat traversal technologies" enum by @ansa89.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: 8 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[toxcore/tox.api.h, line 522 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KX1HmdtYaneO-qO-qi4:-KX3dcEhYxgvowAXRyAG:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/b3eca744e07fd4b0c7ca5aef222c0f7017e25631/toxcore/tox.api.h#L522)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> Replace comma with period.
</details>
Done.

---

*[toxcore/tox.api.h, line 522 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KX-aBa1Z-sUi-Rzz4MF:-KX3ddpsOdry1TQpZtbu:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/b3eca744e07fd4b0c7ca5aef222c0f7017e25631/toxcore/tox.api.h#L522)):*
<details><summary><i>Previously, nurupo wrote…</i></summary>
> Should we be more explicit about how the true/false maps to "enables"/"disables"? Reading this I could guess that setting `nat_level = true` would enable hole-punching, but I don't like to guess.
</details>
Done.

---

*[toxcore/tox.api.h, line 524 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KWw9XFtjn3xu7pL3QwN:-KX3df5Sm1hgncquft_8:b-41kr51) ([raw file](https://github.com/toktok/c-toxcore/blob/b3eca744e07fd4b0c7ca5aef222c0f7017e25631/toxcore/tox.api.h#L524)):*
<details><summary><i>Previously, iphydf wrote…</i></summary>
> For now, we're sure, but anyway this bool is going away very soon, when it becomes part of the "nat traversal technologies" enum by @ansa89.
</details>
renamed, and per IRC conversation, it will eventually be a typedef enum, but not this time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

iphydf commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 7 of 13 files at r6, 2 of 4 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-:-KX3ekrE8DgNdfCHc7r1:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 10 of 12 files reviewed at latest revision, 3 unresolved discussions, some commit checks broke.

---

*[toxcore/Messenger.c, line 1937 at r9](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-KX3iJ5YKwM_IwaHk8x0:-KX3iJ5YKwM_IwaHk8x1:bux9sxc) ([raw file](https://github.com/toktok/c-toxcore/blob/cb65d8c606eb63854938ba96635f5b95ecbad275/toxcore/Messenger.c#L1937)):*
> ```C
>     }
> 
>     m->dht = new_DHT(m->log, m->net, options->nat_level);
> ```

hole_punching_enabled

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Do we need some kind of auto-test for this new functionality?

Otherwise it looks good.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 2 of 2 files at r8, 2 of 2 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/262#-:-KXOYBpSWl131ZDgyUBW:bw4yjf8)*
<!-- Sent from Reviewable.io -->

iphydf commented:
I don't think it's possible to auto-test it.
